File: hdd/merge_two_sorted_list/answer.dart
/**
 * Definition for singly-linked list.
 * class ListNode {
 *   int val;
 *   ListNode? next;
 *   ListNode([this.val = 0, this.next]);
 * }
 */

class Solution {
  ListNode? mergeTwoLists(ListNode? list1, ListNode? list2) {
    // 둘중 하나라도 null이 아닐 때
    if(list1 != null && list2 != null){
      ListNode mergedList = ListNode();
      ListNode? pointer = mergedList;

      while(list1 != null && list2 != null){
        if(list1.val <= list2.val){
          pointer?.next = list1;
          list1 = list1.next;
        } else {
          pointer?.next = list2;
          list2 = list2.next;
        }

        pointer = pointer?.next;
      }

      if(list1 == null){
        pointer?.next = list2;
      } else {
        pointer?.next = list1;
      }
      return mergedList.next;
    } else {
      if(list1 == null){
        return list2;
      }
      return list1;
    }
  }
}
//Runtime: 336 ms, faster than 100.00% of Dart online submissions for Merge Two Sorted Lists.
// Memory Usage: 141.8 MB, less than 100.00% of Dart online submissions for Merge Two Sorted Lists.

File: hdd/merge_two_sorted_list/answer.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun mergeTwoLists(list1: ListNode?, list2: ListNode?): ListNode? {
        var mergedList = ListNode();
        var pointer : ListNode? = mergedList;

        var list1Pointer = list1;
        var list2Pointer = list2;

        while(list1Pointer != null && list2Pointer != null){
            if(list1Pointer.`val` <= list2Pointer.`val`){
                pointer?.next = list1Pointer;
                list1Pointer = list1Pointer?.next;
            } else {
                pointer?.next = list2Pointer;
                list2Pointer = list2Pointer?.next;
            }

            pointer = pointer?.next;
        }

        if(list1Pointer == null){
            pointer?.next = list2Pointer;
        } else {
            pointer?.next = list1Pointer;
        }

        return mergedList.next;
    }
}

// Runtime: 183 ms, faster than 94.19% of Kotlin online submissions for Merge Two Sorted Lists.
// Memory Usage: 35.4 MB, less than 88.84% of Kotlin online submissions for Merge Two Sorted Lists.